Text Editing Model Refactor

justinmc opened the pull request:
This PR aims to improve the separation of concerns of our text editing classes.

Design doc: https://docs.google.com/document/d/1lkpg4dfVZif9NJ_dDQ01QkwQC3OqzhUFahhjgv_rcbM/edit#

Related: https://github.com/flutter/flutter/issues/80226

### Summary

- TextEditingValue and TextSelection now contain simple methods like extendTo, deleteTo, etc. for manipulating themselves.  The logic was abstracted from RenderEditable.
- TextEditingActionTarget now contains the more specific text manipulation methods, which often call out to TextEditingValue (e.g. extendSelectionRightByLine, deleteToStart, etc.).  These were also abstracted from RenderEditable.
- TextMetrics is a new abstract class, implemented by RenderEditable, that exposes an API for getting text position information only.  It's sometimes used by the TextEditingActionTarget methods.
- Overall result: RenderEditable shrinks by ~2,000 lines of logic unrelated to rendering.  All of that logic no longer has access to data from multiple frames, no write access beyond the TextEditingValue.

### Reviewing
Sorry this is such a huge diff!  It's all moving code around and refactoring it; very little totally new.  You can skim most of it because of this.  I tried to leave helpful comments where big chunks of code were moved.  Let me know if I can explain anything.

My commits are also pretty messy as I explored ideas and changed approaches 😁

justinmc committed:
Static methods out of RE and into TEM

justinmc committed:
WIP

justinmc committed:
WIP viable const class?

justinmc committed:
Exploring a simplified TEM, have an idea about TextMetrics from Weiyu in a comment

justinmc committed:
Working expandSelectionLeftByLine using TextMetrics

justinmc committed:
Put expandSelectionLeftByLine field logic in TextEditingActionTarget instead of in DefaultTextEditingActions

justinmc committed:
delete via the new setup

justinmc committed:
moveSelectionLeftByLine works in the new style

justinmc committed:
Moved all selection methods to TEM and target

justinmc committed:
TextEditingModel deleted and everything moved into TextEditingValue

justinmc committed:
Method order

justinmc committed:
Obsolete todos

justinmc committed:
Fix bug caused by me messing with RE parameters and forgetting to put them back...

justinmc commented on packages/flutter/lib/src/services/text_input.dart:
```diff
@@ -653,110 +643,6 @@ class RawFloatingCursorPoint {
 
-/// The current text, selection, and composing state for editing a run of text.
-@immutable
-class TextEditingValue {
```
TextEditingValue was moved to its own file in order to not pollute this one, now that it's much larger.

justinmc commented on packages/flutter/lib/src/widgets/text_editing_action.dart:
```diff
@@ -19,12 +22,847 @@ import 'framework.dart';
+      return _extendSelectionToStart(cause);
+    }
+
+    // TODO(justinmc): I think this assert has to happen in RenderEditable. If
```
This is probably one of the big things I still need to do.  I think that doing the assert in RenderEditable as I described in this comment is doable.  I kind of wish we didn't need this assertion at all, though.

justinmc commented on packages/flutter/lib/src/rendering/editable.dart:
```diff
@@ -77,47 +77,6 @@ class TextSelectionPoint {
   }
 }
 
-// Check if the given code unit is a white space or separator
```
This file's diff is really hard to read, but basically what happened was:

  * Lots of methods moved to TextEditingAction and TextEditingValue.
  * Some alphabetizing/reordering may have happened...
  * TextMetrics was implemented.

justinmc committed:
Remove trailing spaces

justinmc committed:
Move TextMetrics to its own file

justinmc committed:
Fix dependency loop

justinmc committed:
Fix recursive import again

justinmc committed:
WIP update references in tests

justinmc committed:
Simplify TextEditingValue API to just move/expand/extendTo, TextEditingAction handles all the cases

justinmc committed:
Simplify RenderEditable to TextMetrics in TextEditingActionTarget

justinmc committed:
Refactor tests to text_editing_action_test

justinmc committed:
Merge branch 'master' into text-editing-model

justinmc committed:
Refactor expandSelectionTo to fix a bug with when the selection swaps order

justinmc committed:
Moved obvious static methods from TextEditingValue into TextEditingActionTarget

justinmc committed:
Move deleteByWord to TextEditingActionTarget, and leave a simple
deleteTo in TextEditingValue.

justinmc committed:
Port deleteByLine and deleteToEnd to TextEditingActions with a generic deleteTo in TextEditingValue

justinmc committed:
deleteToStart moved

justinmc committed:
deleteForward moved

justinmc committed:
Port deleteForwardByWord

justinmc committed:
deleteForwardByLine ported

justinmc committed:
Port last delete method to TextEditingActions

justinmc committed:
Comment and do setSelection with a default implementation

justinmc committed:
Fix failing test by aligning it with its corresponding TextField test (https://github.com/flutter/flutter/blob/574dde327d252a5d2776e3cc62b063ef3b9e1919/packages/flutter/test/material/text_field_test.dart\#L5415). I assume it was wrong...

justinmc committed:
Fixing lint errors and commented out tests

justinmc committed:
Fix analyzer errors

justinmc committed:
Merge branch 'master' into text-editing-model

justinmc committed:
Fix analyzer errors

justinmc committed:
Fix missing docs references

justinmc committed:
Merge branch 'master' into text-editing-model
Made sure end of line affinity fix still worked and was tested.

justinmc committed:
Linter and docs fixes

justinmc committed:
Fix editable text tests by passing through affinity

justinmc committed:
Fix analyzer

justinmc committed:
Merge branch 'master' into text-editing-model

justinmc committed:
Fix incorrect test

justinmc committed:
Docs todo

justinmc committed:
Consolidate value and textEditingValue

justinmc committed:
TextEditingActionTarget to its own file

justinmc committed:
Bring back the layout assert via an overridable method

justinmc committed:
Only set selection/value if it changed

justinmc committed:
Clean up todo

justinmc committed:
Analyzer fixes, and move TextEditingActionTarget tests to own file.

justinmc committed:
Merge branch 'master' into text-editing-model

justinmc committed:
Analyzer fix

justinmc committed:
Confusion with the TEV aliases and a setter on one. Noted on issue 87748

justinmc committed:
Fix tests due to shorting setSelection/Value in wrong place

justinmc commented on packages/flutter/test/widgets/editable_text_test.dart:
```diff
@@ -4414,7 +4414,7 @@ void main() {
           baseOffset: 20,
           extentOffset: 20,
-          affinity: TextAffinity.downstream,
+          affinity: TextAffinity.upstream,
```
Previously, we were usually resetting the affinity when manipulating the selection.  I updated everything to preserve the affinity, or explicitly set it when it makes sense (like moving the selection to the start should be upstream).  I updated these tests to match.

justinmc commented on packages/flutter/lib/src/services/text_editing_value.dart:
```diff
@@ -0,0 +1,444 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
```
Are we ok with growing the API of TextEditingValue like this?

chunhtai commented on packages/flutter/lib/src/rendering/editable.dart:
```diff
@@ -471,1767 +439,207 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin,
+  /// {@macro flutter.services.TextMetrics.getTextPositionVertical}
+  @override
+  TextPosition getTextPositionVertical(int textOffset, double verticalOffset) {
+    final Offset caretOffset = _textPainter.getOffsetForCaret(TextPosition(offset: textOffset), _caretPrototype);
```
This is where the one frame delay I mentioned in the doc. the _textPainter.text may be out of date from the text editing value. The caretOffset, or any other metrics render editable collected from the textpainter, may be out of date for one frame. I added a hack in _setSelection to get around this problem, but I am not sure whether there will be other corner case

justinmc replied:
Ah I see, thanks for the explanation.  I'll take another look at what we can do to mitigate that problem after this is merged.

chunhtai commented on packages/flutter/lib/src/services/text_editing_value.dart:
```diff
@@ -0,0 +1,444 @@
+  ///
+  /// Setting includeWhitespace to false will only return the index of non-space
+  /// characters.
+  static int nextCharacter(int index, String string, [bool includeWhitespace = true]) {
```
I have written similar typographic methods on the engine side
https://github.com/flutter/engine/blob/e0ebbb226f38e552289e380324943ec67126d325/shell/platform/android/io/flutter/view/AccessibilityBridge.java#L1192
and
https://github.com/flutter/engine/blob/e0ebbb226f38e552289e380324943ec67126d325/shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm#L437

I hope at some point we can use the API from skia to calculate code unit and avoid duplicating the logic in a bunch of places...

justinmc replied:
Ah good catch.  I remember talking to the Dart team about this but can't remember the details.  This seems to work with my testing.

I'll refactor this to use CharacterRange, move it into test_editing_action_target.dart, and make it `visibleForTesting` only.

justinmc replied:
Here's that change: bd27952

chunhtai reviewed (commented):
Is it possible to separate this to smaller code chunks?

LongCatIsLooong commented on packages/flutter/lib/src/services/text_editing_value.dart:
```diff
@@ -0,0 +1,444 @@
+    final int startPoint = includeWhitespace || !isWhitespace(text.codeUnitAt(offset))
+        ? offset
+        : nextCharacter(offset, text, includeWhitespace);
+    final TextRange nextWord = textMetrics.getWordBoundary(TextPosition(offset: startPoint));
```
Ah I see `TextMetrics` acts as a tokenizer here. But I don't think this is right: `textMetrics` may have different text than the `text` passed in.

justinmc replied:
You're right, same problem as ever in that the TextEditingValue and TextPainter's text can be out of sync.  I'll follow up on this.

LongCatIsLooong commented on packages/flutter/lib/src/services/text_editing_value.dart:
```diff
@@ -0,0 +1,444 @@
+    );
+  }
+
+  /// Return [selection] expanded to the given [TextPosition].
```
It's not clear to me what the differences are between expand and extend from reading the documentation.

justinmc replied:
I knew that was going to be confusing.  I'll add some docs about it in both expandSelectionTo and extendSelectionTo.

Basically, "extend" moves the extentOffset while holding the baseOffset fixed, but "expand" grows the selection to include the given position.  If you try cmd+shift+left/right on a Mac, that's expand, while option+shift+up/down is extend.

justinmc replied:
I've improved the docs: https://github.com/flutter/flutter/pull/86736/commits/0675ec4ad718f1202101edf3d0c1d6167a4bd9ce

Would it be more intuitive if I renamed "extend" to "pivot"?

LongCatIsLooong commented on packages/flutter/lib/src/services/text_editing_value.dart:
```diff
@@ -0,0 +1,444 @@
+  /// Select the entire text value.
+  /// {@endtemplate}
+  TextSelection selectAll() {
+    return selection.copyWith(
```
Why the `copyWith`? Is it for keeping the original `isDirectional`? (I'm not sure what `isDirectional` is for, none of our platform text input plugins set that to true it seems).

justinmc replied:
Yes, I didn't want to mess with that as I also wasn't sure where it matters.

justinmc replied:
I ended up removing selectAll and doing it all in TextEditingActionTarget. 6e74a3d

LongCatIsLooong commented on packages/flutter/lib/src/services/text_editing_value.dart:
```diff
@@ -0,0 +1,444 @@
+}
+/// The current text, selection, and composing state for editing a run of text.
+@immutable
+class TextEditingValue {
```
Maybe we should have a `TextEditingValue replace(TextRange, String)` method like Android does? One of the biggest pain points from my personal experience is input manipulation is tricky when you have to take the composing range and selection range into account when you just want to change the text.

justinmc replied:
I'd like to hold off on this to avoid increasing the size of this PR any more if that's ok with you, but I really like the idea.  Lots of the existing code in the delete methods here spend a lot of lines getting the composing region right.  It would be cool to just remove the relevant text from the string and then call `replace` and let it figure out composing and selection.  I'll keep an eye out for other opportunities for this.

LongCatIsLooong replied:
Yeah the android `Editable` implementation implements most common text manipulation methods in terms of `replace`, since it's `replace` implementation takes care of the selection and composition.

justinmc replied:
I investigated adding `replace` to this PR again, but it is making this PR even bigger and more confusing.  I'll follow up in another PR based on my branch https://github.com/justinmc/flutter/tree/text-editing-value-replace.  In the meantime I've removed all changes to TextEditingValue.  `deleteTo` and related stuff is not private inside of TextEditingActionTarget.  In the new PR, I'll replace that with calls to `replace`.

LongCatIsLooong commented on packages/flutter/lib/src/widgets/text_editing_action_target.dart:
```diff
@@ -0,0 +1,1224 @@
+///
+///   * [EditableTextState], which implements this and is the most typical
+///     target of a TextEditingAction.
+abstract class TextEditingActionTarget {
```
This class seems to have way too many methods that are very specific. Hopefully we can distill them down to a handful as basis.

justinmc replied:
That seems like it'll be good to me, I'll ping you for more details later.

I thought about putting all of the logic that is currently in these methods directly into the Action handlers instead, which would also work.  However I worry about encouraging us/developers to use Actions when plain ol' methods are easier to work with and understand.

I have simplified the methods in TextEditingValue per your recommendation.  So TextEditingValue contains simple methods like moveSelectionTo, while TextEditingActionTarget contains the higher level methods like moveSelectionForwardByWord.

LongCatIsLooong reviewed (commented):
Unfortunately I don't think this fixes https://github.com/flutter/flutter/issues/80226 yet (if the issue includes `deleteByWord` and `deleteByLine`, etc). But this looks like a good step towards that!

justinmc committed:
No need for includeWhitespace in delete method

justinmc commented on packages/flutter/lib/src/services/text_editing_value.dart:
```diff
@@ -0,0 +1,444 @@
+
+  /// Return the offset at the end of the nearest word to the right of the given
+  /// offset.
+  static int getRightByWord(String text, TextMetrics textMetrics, int offset, [bool includeWhitespace = true]) {
```
I actually had already moved this to TextEditingActionTarget as a private instance method that uses TextMetrics.getWordPosition, but I had forgotten to delete this static method here.  Thanks for catching this.  Let me know if you like it on TextEditingActionTarget or if you think it would be better on TextMetrics.

Regarding RTL, you're right, I think we have a lot of missing support for this.  I've created an issue (https://github.com/flutter/flutter/issues/88126) and I'll add TODOs for it here.

justinmc committed:
Remove dead getRightByWord, and link to RTL issue on active getRight/LeftByWord methods

justinmc committed:
Use TextPosition where relevant rather than an integer index

justinmc commented on packages/flutter/lib/src/services/text_metrics.dart:
```diff
@@ -0,0 +1,38 @@
+  /// If the offset is already on the first line, the given offset will be
+  /// returned.
+  /// {@endtemplate}
+  TextPosition getTextPositionAbove(int offset);
```
Good catch, I tried to convert most things to use a TextPosition but missed these.

justinmc committed:
expand/extend differences docs

justinmc commented on packages/flutter/lib/src/services/text_editing_value.dart:
```diff
@@ -0,0 +1,444 @@
+    );
+  }
+
+  /// Deletes to the given index.
```
I've changed it to say "inclusive" but tell me if I'm wrong here...

`wo|rd`

deleteTo(0) => `|rd`
deleteTo(4) => `wo|`

Is that right?  I think it confuses me because I'm thinking of the indices as the positions between the characters, not the characters themselves.

LongCatIsLooong replied:
Yeah that's my understanding as well.

justinmc committed:
Specify inclusive

justinmc committed:
Remove TODO for RTL support, as the current behavior is actually consistent with most platforms (see #88126)

justinmc committed:
Analyzer fix extra import

justinmc committed:
Put selection methods on TextSelection?

justinmc commented on packages/flutter/lib/src/services/text_editing_value.dart:
```diff
@@ -0,0 +1,444 @@
+  }
+
+  /// Return [selection] collapsed and moved to the given [TextPosition].
+  TextSelection moveSelectionTo(TextPosition position) {
```
Done here: b3864a9

I feel a little uneasy about adding new methods to TextSelection just like I did originally for TextEditingValue, but maybe there's nothing to worry about.  These methods are pretty fundamental.

justinmc committed:
Merge branch 'master' into text-editing-model

justinmc committed:
Analyzer fixes

justinmc committed:
Fix right aligned position after wordwrap

justinmc committed:
Merge branch 'master' into text-editing-model

justinmc committed:
Analyzer and test fixes

justinmc commented:
@chunhtai @LongCatIsLooong Ready for re-review.

About splitting this up into smaller PRs, unfortunately I think it's more work than it's worth at this point.  Sorry, I know the diff is enormous.  I tried to leave comments where things were moved and refactored so it's not like reviewing brand new code.

About the cross-frame problem, I may take a look at fixing the root part in another PR.  At least all of these methods in this PR have access to one source of truth only now.

Lastly, the only failure now is the typical SVG customer test, which I have fixed before by squashing all the commits.  Hopefully it just goes away before I have to do that.

justinmc committed:
Merge branch 'master' into text-editing-model

LongCatIsLooong commented on packages/flutter/lib/src/services/text_editing_value.dart:
```diff
@@ -0,0 +1,326 @@
+      return _deleteNonEmptySelection();
+    }
+
+    final String textBefore = selection.textBefore(text);
```
There seems to be code duplication, 243-260 looks very similar to 262 - 281. This part of the implementation will probably be cleaner if we first figure out the normalized `TextRange` to delete, and then use "start" and "end" instead of "index" and "selection.extentOffset".

LongCatIsLooong commented on packages/flutter/lib/src/services/text_metrics.dart:
```diff
@@ -0,0 +1,38 @@
+  TextPosition getTextPositionAbove(TextPosition position);
+
+  /// {@template flutter.services.TextMetrics.getTextPositionBelow}
+  /// Returns the TextPosition below the given offset into _plainText.
```
`_plainText` is private.

LongCatIsLooong commented on packages/flutter/lib/src/services/text_editing.dart:
```diff
@@ -144,4 +145,106 @@ class TextSelection extends TextRange {
+  /// to the given [TextPosition].  It doesn't strictly grow the selection and
+  /// may collapse it or flip its order.
+  TextSelection expandTo(TextPosition position, [bool extentAtIndex = false]) {
+    final int upperOffset = math.min(baseOffset, extentOffset);
```
This is `end`. and `lowerOffset` is `start`.

LongCatIsLooong replied:
nit: the implementation is a bit difficult to read.

justinmc replied:
Agreed.  I cleaned this up a bit and added some comments.

justinmc committed:
Code review

justinmc committed:
Code review docs improvement

justinmc committed:
Merge branch 'master' into text-editing-model

justinmc committed:
Simplify deleteTo thanks to code review

justinmc committed:
Simplify expandTo thanks to code review as well

justinmc commented:
@LongCatIsLooong Ready for re-review.

LongCatIsLooong commented on packages/flutter/lib/src/services/text_metrics.dart:
```diff
@@ -0,0 +1,38 @@
+  /// {@template flutter.services.TextMetrics.getLineAtOffset}
+  /// Return a [TextSelection] containing the line of the given [TextPosition].
+  /// {@endtemplate}
+  TextSelection getLineAtOffset(String text, TextPosition position);
```
IMO this is going to be error prone. `TextMetrics`'s layout information is highly dependent on its text, so being able to provide a potentially different String is a bit confusing. 

> enforce what text it's using

Yeah but we can't layout a plain `String`. It has to be styled and you need other inputs for instance the incoming box constraints. Also other methods in the interface don't seem to take a String? All `TextMetrics` implementations are likely going to have states.

And for `RenderEditable.getLineAtOffset`, getting the length of the obscured text should be an implementation detail instead of being exposed as a parameter I think.

justinmc replied:
You're right, this parameter makes the multiple sources of truth thing more confusing, not less.  What we need to do is ensure that the instance of TextMetrics etc. are consistent with the state they're operating on.  I'll remove the parameter and use `_plainText`.

justinmc committed:
deleteTo uses TextPosition instead of index, for consistency with other similar methods

justinmc commented on packages/flutter/lib/src/rendering/editable.dart:
```diff
@@ -419,1820 +378,264 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin,
-  void selectAll(SelectionChangedCause cause) {
-    textSelectionDelegate.selectAll(cause);
+  /// Returns the TextPosition above or below the given offset.
+  TextPosition getTextPositionVertical(TextPosition position, double verticalOffset) {
```
Yup good catch.

justinmc committed:
getPositionVertical can be private

justinmc committed:
Use fromPosition and get rid of moveTo

justinmc committed:
Analyzer fix

justinmc commented on packages/flutter/lib/src/services/text_editing.dart:
```diff
@@ -144,4 +144,97 @@ class TextSelection extends TextRange {
+  }
+
+  /// Return the selection collapsed and moved to the given [TextPosition].
+  TextSelection moveTo(TextPosition position) {
```
Good point, which as we've discussed doesn't seem to be relevant to this.  I'll swap in `fromPosition` and get rid of this method.

justinmc commented on packages/flutter/lib/src/services/text_metrics.dart:
```diff
@@ -0,0 +1,74 @@
+  ///
+  /// Includes newline characters from ASCII and separators from the
+  /// [unicode separator category](https://www.compart.com/en/unicode/category/Zs)
+  static bool isWhitespace(int codeUnit) {
```
It looks like it's used in 3 different files now unfortunately (RenderEditable, TextEditingValue, TextEditingActionTarget).  Probably not worth it to duplicate the method three times.

I see @gspencergoog's comment about this:

```
// TODO(gspencergoog): replace when we expose this ICU information.
```

Any idea if we still have plans to expose that, or if there's a cleaner way for me to expose this one?

gspencergoog replied:
I don't know if we plan to expose it, but I think the problem was that it adds to executable size too much to add the tables in that ICU uses.

justinmc committed:
Dont pass text to TextMetrics, it really should operate on only one internal string

justinmc committed:
Implement next/prevChar with CharacterRange for performance, and make them private

justinmc committed:
Analyzer fixes

justinmc committed:
Remove selectAll method, do it in place

justinmc committed:
Undo changes to TextEditingValue and move to TextEditingActionTarget. replace method will follow in another PR

LongCatIsLooong reviewed (approved):
LGTM. I wonder if there's a better name for `TextMetrics`. `TextLayout`?

justinmc committed:
TextMetrics rename to TextLayoutMetrics
